test/join.jl

```diff
@@ -116,5 +116,204 @@ module TestJoin
+                                         NullableVector{Int}]
+        @test o(on) == DataTable(id = NullableArray([1, 3, 5, N, N, N]),
+                                 fid = NullableArray([1, 3, 5, 0, 2, 4]),
+                                 id_1 = NullableArray([1, 3, N, 0, 2, 4]))
+        @test typeof.(o(on).columns) == [NullableVector{Int},
+                                         NullableVector{Float64},
+                                         NullableVector{Int}]
+
+        on = [:id, :fid]
+        @test i(on) == DataTable(Any[[1, 3], [1, 3]], [:id, :fid])
+        @test typeof.(i(on).columns) == [Vector{Int}, Vector{Float64}]
+        @test l(on) == DataTable(id = NullableArray([1, 3, 5]),
+                                 fid = NullableArray([1, 3, 5]))
+        @test typeof.(l(on).columns) == [NullableVector{Int},
+                                         NullableVector{Float64}]
+        @test r(on) == DataTable(id = NullableArray([1, 3, 0, 2, 4]),
+                                 fid = NullableArray([1, 3, 0, 2, 4]))
+        @test typeof.(r(on).columns) == [NullableVector{Int},
+                                         NullableVector{Float64}]
+        @test o(on) == DataTable(id = NullableArray([1, 3, 5, 0, 2, 4]),
+                                 fid = NullableArray([1, 3, 5, 0, 2, 4]))
+        @test typeof.(o(on).columns) == [NullableVector{Int},
+                                         NullableVector{Float64}]
+    end
+
+    @testset "all joins with CategoricalArrays" begin
+        small = DataTable(Any[CategoricalArray([1, 3, 5]),
+                              CategoricalArray([1.0, 3.0, 5.0])], [:id, :fid])
+        large = DataTable(Any[CategoricalArray([0, 1, 2, 3, 4]),
+                              CategoricalArray([0.0, 1.0, 2.0, 3.0, 4.0])], [:id, :fid])
+        N = Nullable()
+
+        @test join(small, large, kind=:cross) ==
+            DataTable(Any[repeat([1, 3, 5], inner = 5),
+                          repeat([1, 3, 5], inner = 5),
+                          repeat([0, 1, 2, 3, 4], outer = 3),
+                          repeat([0, 1, 2, 3, 4], outer = 3)],
+                      [:id, :fid, :id_1, :fid_1])
+        @test typeof.(join(small, large, kind=:cross).columns) ==
+            [CategoricalVector{i, UInt32} for i in [Int, Float64, Int, Float64]]
+
+        i(on) = join(small, large, on = on, kind = :inner)
+        l(on) = join(small, large, on = on, kind = :left)
+        r(on) = join(small, large, on = on, kind = :right)
+        o(on) = join(small, large, on = on, kind = :outer)
+        s(on) = join(small, large, on = on, kind = :semi)
+        a(on) = join(small, large, on = on, kind = :anti)
+
+        @test s(:id) ==
+              s(:fid) ==
+              s([:id, :fid]) == DataTable(Any[[1, 3], [1, 3]], [:id, :fid])
+        @test typeof.(s(:id).columns) ==
+              typeof.(s(:fid).columns) ==
+              typeof.(s([:id, :fid]).columns) == [CategoricalVector{Int, UInt32},
```

Comment:
Hmm, or replace it with `CategoricalArrays.DefaultRefType` since you need it here.